resample/bootstrap.py

```diff
@@ -0,0 +1,285 @@
+        'student' (also: 't'), 'laplace', 'logistic', 'F' (also: 'f'),
+        'beta', 'gamma', 'log-normal' (also: 'lognorm', 'log-gaussian'),
+        'inverse-gaussian' (also: 'invgauss'), 'pareto', 'poisson'.
+    strata : array-like, optional
+        Stratification labels. Must have the same shape as `sample`. Default is None.
+    random_state : numpy.random.Generator or int, optional
+        Random number generator instance. If an integer is passed, seed the numpy
+        default generator with it. Default is to use `numpy.random.default_rng()`.
+
+    Notes
+    -----
+    Stratification:
+
+    If data is not iid, but consists of several distinct classes, stratification
+    ensures that the relative proportions of each class are maintained in each
+    replicated sample. This is a stricter constraint than that offered by the
+    balanced bootstrap, which only guarantees that classes have the original
+    proportions over all replicas.
+
+    Yields
+    ------
+    ndarray
+        Bootstrap sample.
+    """
+    if random_state is None:
+        rng = np.random.default_rng()
+    elif isinstance(random_state, int):
+        rng = np.random.default_rng(random_state)
+    else:
+        rng = random_state
+
+    sample = np.atleast_1d(sample)
+
+    if strata is not None:
+        strata = np.atleast_1d(strata)
+        if strata.shape != sample.shape:  # type: ignore
+            raise ValueError("a and strata must have the same shape")
+        return _resample_stratified(sample, size, method, strata, rng)
+
+    if method == "balanced":
+        return _resample_balanced(sample, size, rng)
+    if method == "ordinary":
+        return _resample_ordinary(sample, size, rng)
+
+    dist = {
+        # put aliases here
+        "gaussian": stats.norm,
+        "normal": stats.norm,
+        "log-normal": stats.lognorm,
+        "log-gaussian": stats.lognorm,
+        "inverse-gaussian": stats.invgauss,
+        "student": stats.t,
+    }.get(method, None)
+
+    if dist is None:
+        # use scipy.stats name
+        try:
+            dist = getattr(stats, method.lower())
+        except AttributeError:
+            raise ValueError("Invalid family: '{}'".format(method))
+
+    if sample.ndim > 1:
+        if dist != stats.norm:
+            raise ValueError("family '%s' only supports 1D samples" % method)
+        if sample.ndim > 2:
+            raise ValueError("multivariate normal only works with 2D samples")
+        dist = stats.multivariate_normal
+
+    return _resample_parametric(sample, size, dist, rng)
+
+
+def bootstrap(fn: Callable, sample: np.ndarray, size: int = 100, **kwds) -> np.ndarray:
+    """
+    Calculate function values from bootstrap samples.
+
+    Generator of bootstrap samples.
+
+    Parameters
+    ----------
+    fn : Callable
+        Bootstrap samples are passed to this function.
+    sample : array-like
+        Original sample.
+    **kwds
+        Keywords are forwarded to `resample`.
```

Comment:
Again I feel this might be unnecessary cognitive overhead for a user, who now has to think in terms of two functions / signatures instead of the one they're using

Reply:
I hope I can convince you otherwise.

Reply:
As explained above, I think that the resample generator is the core of the library, not the `bootstrap` or the `jackknife` functions. Instead of forwarding the arguments to the resampler, we could also pass the generator instance to those methods. With the generator factorized out, the functions `jackknife` and `bootstrap` essentially do the same thing.

Reply:
Is there a benefit to passing them as kwargs instead of explicitly, aside from coding convenience? Personal preference but I find kwargs a bit opaque in general and usually would rather clearly enumerate what a function needs in order to work.

Reply:
The benefit is that there is no redundancy. Redundancy in code is bad:
https://de.wikipedia.org/wiki/Don%E2%80%99t_repeat_yourself.

This is not about coding convenience, I design interfaces with the user in mind, but there are always trade-offs. If there is no redundancy, there is no possibility of information going out of sync. There is no possibility for keywords to differ in `resample` and the other methods that use it. The burden for the user to look up another function is small, and offset by the fact that the total documentation gets more compact, because the same information is not repeated.

There is another design, which avoids this problem and is even more elegant:
`def bootstrap(fn, resampler): ...`

Instead of passing the arguments for the `resample` generator, we pass an instance of the `resample` generator. There is no redundancy then in the docs and the interface is clean and simple.

Reply:
We can leave it this way but I'd rather use the convention of calling them `kwargs` instead of `kwds` (can be a follow-up)